### PR TITLE
Handle org level authorization

### PR DESCRIPTION
**Pull request by AnuradhaSK**
### Proposed changes in this pull request
$Subject

### Depends on

- [x] https://github.com/wso2-extensions/identity-organization-management/pull/51

### CODE REVIEW

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java [hunk 1]
```diff
@@ -67,6 +68,18 @@ public void invoke(Request request, Response response) throws IOException, Servl
                 authorizationContext.setIsCrossTenantAllowed(resourceConfig.isCrossTenantAllowed());
                 authorizationContext.setAllowedTenants(resourceConfig.getCrossAccessAllowedTenants());
             }
+
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java:71]: TODO: This should be invoked only if the request path is /o/<org-id>.

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java [hunk 2]
```diff
@@ -67,6 +68,22 @@ public void invoke(Request request, Response response) throws IOException, Servl
...
+            String requestURI = request.getRequestURI();
+            if (requestURI.startsWith("/o/")) {
+                AuthorizationResult authorizationResult =
+                        authorizeInOrganizationLevel(request, response, authenticationContext, resourceConfig,
+                                authorizationContext);
+                /*
+                If the user has organization level access, authorize and proceed.
+                Else evaluate authorization through old authz model.
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java:79]: Is this comment correct?

Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java]: rephrase the comment in [769749c](https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/184/commits/769749cc0c334b7232150455cfb08718ff812631)
Fallback to old authz model is already added as a comment is the below line.

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java [hunk 3]
```diff
@@ -120,6 +136,51 @@ public void invoke(Request request, Response response) throws IOException, Servl
...
+
+        if (!isUserEmpty(authenticationContext)) {
+            String httpMethod = request.getMethod();
+            String tenantDomainFromURLMapping = Utils.getTenantDomainFromURLMapping(request);
+
+            OrganizationManagementAuthorizationContext orgMgtAuthorizationContext =
+                    new OrganizationManagementAuthorizationContext();
+
+            if (resourceConfig != null) {
+                if(StringUtils.isNotEmpty(resourceConfig.getPermissions())) {
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java:154]:
```suggestion
                if (StringUtils.isNotEmpty(resourceConfig.getPermissions())) {
```

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java [hunk 4]
```diff
@@ -120,6 +136,51 @@ public void invoke(Request request, Response response) throws IOException, Servl
...
+
+            List<AuthorizationManager> authorizationManagerList = AuthorizationValveServiceHolder.getInstance()
+                    .getAuthorizationManagerList();
+            AuthorizationManager authorizationManager = HandlerManager.getInstance()
+                    .getFirstPriorityHandler(authorizationManagerList, true);
+            try {
+                authorizationResult = authorizationManager.authorize(orgMgtAuthorizationContext);
+            } catch (AuthzServiceServerException e) {
+                APIErrorResponseHandler.handleErrorResponse(authenticationContext, response,
+                        HttpServletResponse.SC_BAD_REQUEST, null);
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java:178]: is `HttpServletResponse.SC_BAD_REQUEST` the correct status code here?

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java (hunk 3 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java]: fixed 04d1f14

File: components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java (hunk 4 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.authz.valve/src/main/java/org/wso2/carbon/identity/authz/valve/AuthorizationValve.java]: fixed in [aea691b](https://github.com/wso2-extensions/identity-carbon-auth-rest/pull/184/commits/aea691baa6bd686a7f906d6ae3785f847309c754)

**Review by madurangasiriwardena — APPROVED** [review]

**Pull request merged**